### PR TITLE
protocol parser: rdp

### PR DESCRIPTION
**Pull request by lmzke**
Continuation of #4131.

Initial implementation of feature [2314](https://redmine.openinfosecfoundation.org/issues/2314):
1. Add protocol parser for RDP
2. Add transactions for RDP negotiation
3. Add eve logging of transactions

This PR has incorporated all previous reviewer suggestions.  New contributor, all feedback welcome.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Sample output, based on [wireshark RDP-002.pcap](https://wiki.wireshark.org/SampleCaptures?action=AttachFile&do=view&target=RDP-002.pcap.gz):

```json
"rdp": {
  "tx_id": 0,
  "event_type": "initial_request",
  "cookie": "A70067"
}

"rdp": {
  "tx_id": 1,
  "event_type": "initial_response"
}

"rdp": {
  "tx_id": 2,
  "event_type": "connect_request",
  "client": {
    "version": "v5",
    "desktop_width": 1152,
    "desktop_height": 864,
    "color_depth": 15,
    "keyboard_layout": "en-US",
    "build": "Windows XP",
    "client_name": "ISD2-KM84178",
    "keyboard_type": "enhanced",
    "function_keys": 12,
    "product_id": 1,
    "capabilities": [
      "support_errinfo_pdf"
    ],
    "id": "55274-OEM-0011903-00107"
  },
  "channels": [
    "rdpdr",
    "cliprdr",
    "rdpsnd"
  ]
}

"rdp": {
  "tx_id": 3,
  "event_type": "connect_response"
}
```

### CODE REVIEW

File: rust/src/rdp/parser.rs [hunk 1]
```diff
@@ -0,0 +1,1544 @@
...
+/// rdp-spec, section 2.2.1.3.2 Client Core Data
+fn desktop_scale_to_opt(x: u32) -> Option<u32> {
+    if x >= 100 && x <= 500 {
+        Some(x)
+    } else {
+        None
+    }
+}
+
+/// constrains device scale to a range, per spec
```
Comment (jlucovsky) [inline rust/src/rdp/parser.rs]: nit: the scale is constrained to a list of valid values, not a range.

File: src/output-json-rdp.c [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    AppLayerParserRegisterLogger(IPPROTO_TCP, ALPROTO_RDP);
+
+    SCLogDebug("rdp log sub-module initialized.");
+
+    result.ctx = output_ctx;
+    result.ok = true;
+    return result;
+}
+
+#define OUTPUT_BUFFER_SIZE 65535
```
Comment (jlucovsky) [inline src/output-json-rdp.c]: Use `JSON_OUTPUT_BUFFER_SIZE` from `output-json.h` instead.

File: src/output-json-rdp.c [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+static TmEcode JsonRdpLogThreadInit(ThreadVars *t, const void *initdata, void **data)
+{
+    LogRdpLogThread *thread = SCCalloc(1, sizeof(*thread));
+    if (unlikely(thread == NULL)) {
+        return TM_ECODE_FAILED;
+    }
+
+    if (initdata == NULL) {
```
Comment (jlucovsky) [inline src/output-json-rdp.c]: Suggest making this check first

**Comment by jasonish** [comment]
@lmzke We'll address these issues ourselves to get this into master a.s.a.p.  Thanks for the work!

BTW, while we plan to take it in, it will be disabled by default.

**Comment by lmzke** [comment]
@jasonish sounds good.  I am available same-day/today if it would lighten your workload to have me incorporate feedback in this PR, and create a new one.  Just let me know by comment here if so, otherwise I am standing down.

I appreciate all the review cycles from the team.

File: rust/src/rdp/parser.rs (hunk 1 above)
Comment (lmzke) [reply rust/src/rdp/parser.rs]: Good catch, comment as written does not apply.

**Comment by jasonish** [comment]
@lmzke Thanks for the offer.  I think we're OK for this initial one, but will pass further stuff on to you.

By chance do you have any suricata-verify tests?

**Comment by lmzke** [comment]
I only have unit tests, included in the contribution.

For public domain PCAPs, if it helps, I have on occasion used [wireshark](https://wiki.wireshark.org/RDP) and [FreeRDP](https://github.com/FreeRDP/FreeRDP/wiki/Network-Level-Authentication).

**Comment by jasonish** [comment]
Closing. Continued at https://github.com/OISF/suricata/pull/4184

**Pull request closed without merging**